onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlDFT.h

```diff
@@ -384,27 +301,120 @@
             outputTensor->GetDataInterface(outputUnknown.GetAddressOf());
             outputUnknown.As(&outputResource);
 
-            auto isPowerOfTwo = [](uint32_t n) { return (n != 0) && ((n & (n - 1)) == 0); };
-            if (isPowerOfTwo(m_inputDims[m_axis]))
+            // { before_dft_axis, axis, after_dft_axis, real_or_complex }
+            std::array<uint32_t, 4> reshapedInputSize = { 1, 1, 1, inputDims.back() };
+            std::array<uint32_t, 4> reshapedOutputSize = { 1, 1, 1, outputDims.back() };
+
+            size_t reshapedIndex = 0;
+            for (int i = 0; i < inputDims.size() - 1; i++)
             {
-                StockhamFFT(inputResource.Get(), outputResource.Get(), commandList.Get());
+                if (i == m_axis || i == (m_axis + 1))
+                {
+                    reshapedIndex++;
+                }
+                reshapedInputSize[reshapedIndex] *= inputDims[i];
+                reshapedOutputSize[reshapedIndex] *= outputDims[i];
             }
-            else {
-                BluesteinZChirp(inputResource.Get(), outputResource.Get(), commandList.Get());
+
+            auto temporarySize = reshapedInputSize;
+            temporarySize.back() = reshapedOutputSize.back();
+            auto temporaryBufferByteSize = sizeof(float) * std::accumulate(temporarySize.begin(), temporarySize.end(), 1, std::multiplies());
+
+            // Calculate elements and strides
+            std::array<uint32_t, 4> reshapedInputStrides = { 1, 1, 1, 1 };
+            std::array<uint32_t, 4> reshapedOutputStrides = { 1, 1, 1, 1 };
+            std::array<uint32_t, 4> temporaryStrides = { 1, 1, 1, 1 };
+            for (int i = static_cast<int>(reshapedInputSize.size()) - 2; i >= 0; i--)
+            {
+                reshapedInputStrides[i] = reshapedInputSize[i + 1] * reshapedInputStrides[i + 1];
+                reshapedOutputStrides[i] = reshapedOutputSize[i + 1] * reshapedOutputStrides[i + 1];
+                temporaryStrides[i] = temporarySize[i + 1] * temporaryStrides[i + 1];
             }
-            return S_OK;
-        }
-        catch (...)
-        {
-            return E_FAIL;
+
+            // Calculate passes
+            params.StockhamParams.NumberOfPasses = static_cast<unsigned>(log2(params.DFTLength));
```

Comment:
>log2 [](http://example.com/codeflow?start=73&length=4)

Double checking... if someone passes 255, do we want that to be 8 passes or 7? log2 will return 7.999 which gets rounded down, but maybe we need 8 passes via `ceil()`? Granted, I read we don't support non-powers of two currently anyway, but I'd like to know which is correct.

Reply:
Yeah, the length needs to be a power of 2. Otherwise fft radix 2 just explodes and cant be used anyway.